numpy/lib/function_base.py

```diff
@@ -3196,8 +3189,16 @@ def i0(x):
 
     """
     x = np.asanyarray(x)
-    x = np.abs(x)
-    return piecewise(x, [x <= 8.0], [_i0_1, _i0_2])
```

Comment:
Wow, sorry that my #11233 missed this. That said, I think the implementation is clearer as I had it. Would the following work:
```
x = np.asanyarray(x)
return piecewise(x, [np.abs(x) <= 8.0], , [_i0_1, _i0_2])
```

Reply:
Oops, sorry, should be
```
x = np.asanyarray(x)
absx = np.abs(x)
return piecewise(x if x.dtype.kind == 'c' else absx, [absx <= 8.0], , [_i0_1, _i0_2])
```

Reply:
But then it won't promote types.

Reply:
I think it will - at least I just checked with my second suggestion, and all tests pass.

Reply:
That looks wrong to me - why take the absolute value of reals but not complex? That makes `-1` and `-1+0j` give different results.

Reply:
I think that my example just makes it more explicit; in both cases:
```
In [2]: np.i0(-1.)
Out[2]: array(1.26606588)

In [3]: np.i0(-1+0j)
Out[3]: array(1.26606588+0.j)
```

Note that `scipy.special.i0` simply doesn't implement complex.

Reply:
Does `return piecewise(x, [absx <= 8.0], , [_i0_1, _i0_2])` not also work?

Reply:
But my trial also revealed what I now realize @peterbell10 was getting at: my suggestion doesn't work properly for integer, but would need a cast of x for the non-complex case. Given that, I'm not sure I can argue any more that it has an advantage of clarify!

Reply:
Really `piecewise` should have a `dtype` argument, but naturally we backed ourself into an API corner that makes it impossible to add...

Reply:
@eric-wieser - one does need `abs(x)` for the float case since otherwise one gets an error in `sqrt` for negative numbers.

Reply:
At the cost of an up to 2x increase in memory usage, we could fix piecewise to call `common_type` internally - I doubt anyone would be upset.

Reply:
@mhvk: see my comment below: that means we handle `complex(-real)` inconsistently with how we handle `-real`.